### PR TITLE
[networks] Reduce HTTP monitoring footprint

### PR DESCRIPTION
**Pull request by p-lambert**
### What does this PR do?

Refactor some bits of the HTTP monitoring code in order to reduce its overhead.
I tested these changes against a load test server processing ~12k requests/second and saw a ~40% reduction in memory usage and a ~10% decrease in CPU.

- [x] Reduce allocations from the HTTP hot path;
- [x] Flatten HTTP map;
- [x] Remove unnecessary mutex from the main event loop;
- [x] Avoid creation of sketches for the single latency samples;
- [x] Intern HTTP path strings;
- [x]  Add hard limit for number of HTTP map entries;

benchmark

Requires https://github.com/DataDog/agent-payload/pull/93

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

Write there any instructions and details that should be tested during the QA.

### CODE REVIEW

File: pkg/network/http/http_stats.go [hunk 1]
```diff
@@ -26,53 +45,87 @@ type RequestStats [5]struct {
...
+// newStats is kept as it is, while the method receiver gets mutated
 func (r *RequestStats) CombineWith(newStats RequestStats) {
-	for i := 0; i < 5; i++ {
-		r[i].count += newStats[i].count
+	for i := 0; i < len(r); i++ {
+		statusClass := 100 * (i + 1)
+		switch newStats[i].Count {
+		case 0:
+			// No data to be merged
+			continue
```
Comment (hmahmood) [inline pkg/network/http/http_stats.go:66]: `continue` is redundant

File: pkg/network/http/http_stats.go [hunk 2]
```diff
@@ -26,53 +45,87 @@ type RequestStats [5]struct {
...
-
 // AddRequest takes information about a HTTP transaction and adds it to the request stats
 func (r *RequestStats) AddRequest(statusClass int, latency float64) {
 	i := statusClass/100 - 1
-	r[i].count++
+	if i >= len(r) {
+		return
+	}
+
+	r[i].Count++
```
Comment (hmahmood) [inline pkg/network/http/http_stats.go:103]: Should this still be incremented if `initSketch` below fails?

File: pkg/network/http/http_statkeeper.go [hunk 3]
```diff
@@ -4,51 +4,63 @@ package http
...
+		SrcPort:   uint16(tx.tup.sport),
+		DstIPHigh: uint64(tx.tup.daddr_h),
+		DstIPLow:  uint64(tx.tup.daddr_l),
+		DstPort:   uint16(tx.tup.dport),
+		Path:      pathString,
+	}
+}
+
+func (h *httpStatKeeper) intern(b []byte) string {
+	v, ok := h.interned[string(b)]
```
Comment (leeavital) [inline pkg/network/http/http_statkeeper.go:60]: I know there were some optimizations to have this not allocate in recent versions of go -- does this string conversion wind up allocating? Since `b` is usually immutable it would be nice to do the map lookup without allocating, and only allocate when we need to intern a new string.

File: pkg/network/http/http_stats.go [hunk 4]
```diff
@@ -26,53 +45,87 @@ type RequestStats [5]struct {
...
-			r[i].latencies = newStats[i].latencies
-		} else if newStats[i].latencies != nil {
-			err := r[i].latencies.MergeWith(newStats[i].latencies)
+			// Finally merge both sketches
+			r[i].Count += newStats[i].Count
+			err := r[i].Latencies.MergeWith(newStats[i].Latencies)
 			if err != nil {
-				log.Debugf("Error merging HTTP transactions: %v", err)
+				log.Debugf("error merging http transactions: %v", err)
 			}
```
Comment (leeavital) [inline pkg/network/http/http_stats.go:91]: I know this isn't new code, but when would we expect a failure when merging two sketches?

File: pkg/network/http/http_stats.go [hunk 5]
```diff
@@ -26,53 +45,87 @@ type RequestStats [5]struct {
...
-	count     int
-	latencies *ddsketch.DDSketch
+	Count     int
+	Latencies *ddsketch.DDSketch
+
+	// This field holds the value (in milliseconds) of the first HTTP request
+	// in this bucket. We do this as optimization to avoid creating sketches with
+	// a single value. This is quite common in the context of HTTP requests without
+	// keep-alives where a short-lived TCP connection is used for a single request.
+	FirstLatencySample float64
```
Comment (leeavital) [inline pkg/network/http/http_stats.go:55]: Is it worth implementing a hacky union type where we only store one of `Latencies` or `FirstLatencySample` in an `interface{}` and use `Count` to assume the type ?

Comment (p-lambert) [reply pkg/network/http/http_stats.go:55]: humm, I honestly don't see a lot of benefit in doing that. I think if we are trying to save 8 bytes using an interface would defeat the purpose of this change since the interface value itself has 2 words already.

File: pkg/network/http/http_statkeeper.go (hunk 3 above)
Comment (p-lambert) [reply pkg/network/http/http_statkeeper.go:60]: Yup, there is an optimization in Golang for this type of lookup: https://github.com/golang/go/commit/f5f5a8b6209f84961687d993b93ea0d397f5d5bf

File: pkg/network/encoding/encoding.go [hunk 6]
```diff
@@ -50,9 +51,11 @@ func modelConnections(conns *network.Connections) *model.Connections {
 	agentConns := make([]*model.Connection, len(conns.Conns))
 	domainSet := make(map[string]int)
 	routeIndex := make(map[string]RouteIdx)
+	httpIndex := FormatHTTPStats(conns.HTTP)
 
 	for i, conn := range conns.Conns {
-		agentConns[i] = FormatConnection(conn, domainSet, routeIndex)
+		httpKey := http.NewKey(conn.Source, conn.Dest, conn.SPort, conn.DPort, "")
+		agentConns[i] = FormatConnection(conn, domainSet, routeIndex, httpIndex[httpKey])
```
[the remark this answers is not in the document]
Comment (p-lambert) [reply pkg/network/encoding/encoding.go:58]: That's a good point. I _think_ it's okay to leave the code as it is because we definitely don't want to "bind" the same stats to multiple connections which would result in counting the same thing multiple times. 

On the other hand, it means that we might be misrepresenting which connections are responsible for which stats.
The obvious fix I guess would be adding PID to the HTTP key (for the record we don't consider PID in DNS either), but I can't think of a realistic scenario where this would be an issue.

One example I can think of is multiple containers in the same kubernetes pod. They share the same network namespace, so they would have the same IP. But I guess the only possible example of tuple collision would be in the case where multiple containers end up using the same source port number at different times within a 30s window which I think it's a pretty contrived scenario.

The other example I guess is forked processes using the same socket, but in that case I guess we don't care about PID resolution since they will be aggregated anyways?

File: pkg/network/encoding/format.go [hunk 7]
```diff
@@ -129,6 +129,40 @@ func FormatCompilationTelemetry(telByAsset map[string]network.RuntimeCompilation
...
+		key.Path = ""
+
+		statsByPath := aggregated[key]
+		if statsByPath == nil {
+			statsByPath = make(map[string]*model.HTTPStats)
+			aggregated[key] = statsByPath
+		}
+
+		ms := &model.HTTPStats{
+			StatsByResponseStatus: make([]*model.HTTPStats_Data, 5),
```
[the remark this answers is not in the document]
Comment (p-lambert) [reply pkg/network/encoding/format.go:145]: The number of status code "buckets" (1xx, 2xx, 3xx, 4xx, 5xx). I should probably extract this to a constant?

File: pkg/network/http/http_stats.go (hunk 4 above)
Comment (p-lambert) [reply pkg/network/http/http_stats.go:91]: err, I don't know 😬 
I'll follow up on that

File: pkg/network/encoding/format.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (hmahmood) [reply pkg/network/encoding/format.go:145]: Yes please

File: pkg/network/http/http_stats.go (hunk 4 above)
Comment (ISauve) [reply pkg/network/http/http_stats.go:91]: MergeWith can return an error if the index mappings don't match: https://github.com/DataDog/sketches-go/blob/c9a8f6a89b352df76f2e49e07d977a9ce163d560/ddsketch/ddsketch.go#L184-L189

This _shouldn't_ happen, but since it's technically possible I added the error handling just do be safe 🤷

File: pkg/network/encoding/encoding.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (leeavital) [reply pkg/network/encoding/encoding.go:58]: Since these metrics contain counts, we'll wind up double counting in the forked process case. 


Could we get around this by adding a `delete(httpIndex, httpKey)` at each iteration?

[the remark this answers is not in the document]
Comment (p-lambert) [reply pkg/network/encoding/encoding.go:58]: oh, you're right! yes, let's do that

File: pkg/network/http/http_stats.go (hunk 2 above)
Comment (p-lambert) [reply pkg/network/http/http_stats.go:111]: yes, I think it's fine to leave like that

**Review by hmahmood — APPROVED** [review]

File: pkg/network/config/config.go [hunk 8]
```diff
@@ -156,7 +156,7 @@ func NewDefaultConfig() *Config {
 		MaxClosedConnectionsBuffered: 50000,
 		MaxConnectionsStateBuffered:  75000,
 		MaxDNSStatsBufferred:         75000,
-		MaxHTTPStatsBuffered:         75000,
+		MaxHTTPStatsBuffered:         100000,
```
Comment (leeavital) [inline pkg/network/config/config.go:159]: Why this change?

**Review by leeavital — APPROVED** [review]

File: pkg/network/config/config.go (hunk 8 above)
Comment (p-lambert) [reply pkg/network/config/config.go:159]: I was trying to get to a number where 1) we wouldn't drop stats for a reasonable request rate 2) result in an acceptable memory usage

In my tests this number is able to process ~3500 requests/s staying around 350MiB
